OpenRA.Mods.Common/Effects/Missile.cs

```diff
@@ -141,52 +200,106 @@ public void Tick(World world)
 
-			// Missile tracks target
+			var cell = world.Map.CellContaining(pos);
+			var terrainHeight = world.Map.MapHeight.Value[world.Map.CellContaining(pos)] * 512;
+
+			// Switch from freefall mode to homing mode
+			if (ticks == info.ActivationDelay + 1)
+			{
+				activated = true;
+				hFacing = OpenRA.Traits.Util.GetFacing(velocity, hFacing);
+			}
+
+			// Switch from homing mode to freefall mode
+			if (info.RangeLimit != 0 && ticks == info.RangeLimit + 1)
+			{
+				activated = false;
+				velocity = new WVec(WDist.Zero, -info.Speed, WDist.Zero)
+					.Rotate(new WRot(WAngle.FromFacing(vFacing), WAngle.Zero, WAngle.Zero))
+					.Rotate(new WRot(WAngle.Zero, WAngle.Zero, WAngle.FromFacing(hFacing)));
+			}
+
+			// Check if target position should be updated (actor visible & locked on)
 			if (args.GuidedTarget.IsValidFor(args.SourceActor) && lockOn)
 				targetPosition = args.GuidedTarget.CenterPosition;
 
+			// Compute current distance from target position
 			var dist = targetPosition + offset - pos;
-			var desiredFacing = OpenRA.Traits.Util.GetFacing(dist, facing);
-			var desiredAltitude = targetPosition.Z;
-			var jammed = info.Jammable && world.ActorsWithTrait<JamsMissiles>().Any(JammedBy);
 
-			if (jammed)
+			WVec move;
+			if (activated)
 			{
-				desiredFacing = facing + world.SharedRandom.Next(-20, 21);
-				desiredAltitude = world.SharedRandom.Next(-43, 86);
+				// If target is not within range and cruise altitudes are used,
+				// aim at the same target but at cruise altitude above current terrain height.
+
+				// Target is considered in range if it would take the missile more time to
+				// travel the distance dist.HorizontalLength at speed info.Speed than to
+				// reach ground using maximum vertical turning downwards starting with the
+				// current vertical facing.
+
+				// Basically vDist is the representation in the x-y plane
+				// of the projection of dist in the z-hDist plane,
+				// where hDist is the projection of dist in the x-y plane.
+
+				// This allows applying vertical rate of turn in the same way as the
+				// horizontal rate of turn is applied.
+				WVec vDist;
+				if (info.CruiseAltitude.Length > 0 && dist.HorizontalLength * info.VerticalRateOfTurn > info.Speed.Length * ((vFacing + 0x40) & 0xFF))
+					vDist = new WVec(-(dist.Z - targetPosition.Z + info.CruiseAltitude.Length + terrainHeight), -dist.HorizontalLength, 0);
+				else
+					vDist = new WVec(-dist.Z, -dist.HorizontalLength, 0);
+
+				// Compute which direction the projectile should be facing
+				var desiredHFacing = OpenRA.Traits.Util.GetFacing(dist, hFacing);
+				var desiredVFacing = OpenRA.Traits.Util.GetFacing(vDist, vFacing);
+
+				// Check whether the homing mechanism is jammed
+				var jammed = info.Jammable && world.ActorsWithTrait<JamsMissiles>().Any(JammedBy);
+				if (jammed)
+				{
+					desiredHFacing = hFacing + world.SharedRandom.Next(-20, 21);
+					desiredVFacing = vFacing + world.SharedRandom.Next(-20, 21);
```

Comment:
It's jamming the *current path* instead of the original path. This gives a compounding jamming effect on the path of the missile.